### PR TITLE
New master version - new features and bug fixes

### PR DESCRIPTION
**Pull request by daanraman**
- New base image for docker (Python instead of Debian)
- Support for "process_documents_chronologically" option on use cases
- Use coefficient of variation instead of standard deviation as metric to detect beaconing
- Print a summary of analysis after outliers has completed
- Improved whitelist processing (better performance & no unnecessary runs)
- Sentry now receives full stack traces instead of an error message when a model fails to run for some reason
- Lots of new unit tests with first set of unit tests to test accuracy of models
- Update elasticsearch python package
- Lots of minor improvements (performance, bugfixes, typos, etc.)

### CODE REVIEW

File: app/helpers/outlier.py [hunk 1]
```diff
@@ -1,52 +1,94 @@
...
+                helpers.singletons.settings.whitelist_literals_config:
             whitelist_values_to_check = each_whitelist_configuration_file_value.split(",")
 
             total_whitelisted_fields_to_match = len(whitelist_values_to_check)
             total_whitelisted_fields_matched = 0
 
             for whitelist_val_to_check in whitelist_values_to_check:
-                if self.matches_specific_whitelist_item_literally(whitelist_val_to_check,
-                                                                  additional_dict_values_to_check):
+                if Outlier.dictionary_matches_specific_whitelist_item_literally(whitelist_val_to_check,
```
Comment (michielmeersmans) [inline app/helpers/outlier.py]: Seems like quite a huge method name

File: app/helpers/outlier.py [hunk 2]
```diff
@@ -55,62 +97,18 @@ def is_whitelisted(self, additional_dict_values_to_check=None):
...
-        _str += "=======\n"
-        _str += "outlier\n"
-        _str += "=======\n"
-
-        for key, value in self.outlier_dict.items():
-            _str += (str(key) + "\t -> " + str(value) + "\n")
-
-        return _str
+    @staticmethod
+    def dictionary_matches_specific_whitelist_item_regexp(regex, set_of_values_to_check):
```
Comment (michielmeersmans) [inline app/helpers/outlier.py]: Method name size

File: requirements.txt [hunk 3]
```diff
@@ -5,7 +5,7 @@ chardet==3.0.4
 configparser==3.5.0
 dateparser==0.6.0
 docopt==0.6.2
-elasticsearch==6.2.0
+elasticsearch==6.4.0
```
Comment (michielmeersmans) [inline requirements.txt]: We're using ELK 6.7 atm, does this exist as python lib?

File: app/tests/unit_tests/test_analyzer_beaconing.py [hunk 4]
```diff
@@ -172,3 +215,39 @@ def test_evaluate_model_beaconing_simple_case(self):
 
         result = [elem for elem in es.scan()][0]
         self.assertEqual(result, expected_doc)
+
+    def _test_whitelist_batch_document_not_process_all(self):  # TODO FIX with new whitelist system
```
[the remark this answers is not in the document]
Comment (daanraman) [reply app/tests/unit_tests/test_analyzer_beaconing.py]: Yes indeed, these are tests being prepared for new improvements not implemented yet.

File: app/tests/unit_tests/test_analyzer_beaconing.py [hunk 5]
```diff
@@ -172,3 +215,39 @@ def test_evaluate_model_beaconing_simple_case(self):
...
+        self.test_es.add_doc(doc2_without_outlier)
+        # Not whitelisted
+        doc3_without_outlier = copy.deepcopy(doc_without_outliers_test_whitelist_03_test_file)
+        self.test_es.add_doc(doc3_without_outlier)
+
+        analyzer.evaluate_model()
+
+        self.assertEqual(len(analyzer.outliers), 2)
+
+    def _test_whitelist_batch_document_no_whitelist_document(self):  # TODO FIX with new whitelist system
```
[the remark this answers is not in the document]
Comment (daanraman) [reply app/tests/unit_tests/test_analyzer_beaconing.py]: Yes indeed, these are tests being prepared for new improvements not implemented yet.

File: app/tests/unit_tests/test_analyzer_metrics.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    def setUp(self):
+        self.test_es = TestStubEs()
+        self.test_settings = TestSettings()
+
+    def tearDown(self):
+        # restore the default configuration file so we don't influence other unit tests that use the settings singleton
+        self.test_settings.restore_default_configuration_path()
+        self.test_es.restore_es()
+
+    def _test_whitelist_batch_document_not_process_all(self):  # TODO FIX with new whitelist system
```
[the remark this answers is not in the document]
Comment (daanraman) [reply app/tests/unit_tests/test_analyzer_metrics.py]: Yes indeed, these are tests being prepared for new improvements not implemented yet.

File: app/tests/unit_tests/test_analyzer_terms.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    def setUp(self):
+        self.test_es = TestStubEs()
+        self.test_settings = TestSettings()
+
+    def tearDown(self):
+        # restore the default configuration file so we don't influence other unit tests that use the settings singleton
+        self.test_settings.restore_default_configuration_path()
+        self.test_es.restore_es()
+
+    def _test_whitelist_batch_document_not_process_all(self):  # TODO FIX with new whitelist system
```
[the remark this answers is not in the document]
Comment (daanraman) [reply app/tests/unit_tests/test_analyzer_terms.py]: Yes indeed, these are tests being prepared for new improvements not implemented yet.

File: app/analyzers/word2vec.py [hunk 8]
```diff
@@ -70,7 +69,7 @@ def evaluate_model(self):
             return
 
         w2v_model = word2vec.Word2Vec(name=self.model_name)
-        search_query = es.filter_by_query_string(self.model_settings["es_query_filter"])
+        search_query = self.search_query
```
[the remark this answers is not in the document]
Comment (daanraman) [reply app/analyzers/word2vec.py]: Fixed

File: app/helpers/outlier.py (hunk 1 above)
Comment (daanraman) [reply app/helpers/outlier.py]: agreed but it does explain what it does so I will keep it

File: requirements.txt (hunk 3 above)
Comment (daanraman) [reply requirements.txt]: 6.4.0 appears to be the latest version https://pypi.org/project/elasticsearch/#history

**Pull request merged**